autoscale/src/main/java/com/sequenceiq/periscope/service/YarnInvocationService.java

```diff
@@ -0,0 +1,138 @@
+
+    private Cluster cluster;
+
+    private LoadAlert loadAlert;
+
+    private LoadAlertConfiguration loadAlertConfiguration;
+
+    private String policyHostGroup;
+
+    public void pollYarnMetricsAndScaleCluster(Long clusterId, String pollingUserCrn, boolean stopStartEnabled, StackV4Response stackV4Response)
+            throws Exception {
+
+        cluster = clusterService.findById(clusterId);
+        loadAlert = cluster.getLoadAlerts().iterator().next();
+        policyHostGroup = loadAlert.getScalingPolicy().getHostGroup();
+        loadAlertConfiguration = loadAlert.getLoadAlertConfiguration();
+
+        Map<String, String> hostFqdnsToInstanceId = stackResponseUtils.getCloudInstanceIdsForHostGroup(stackV4Response, policyHostGroup);
+        List<String> stoppedHostInstanceIds = stackResponseUtils.getStoppedCloudInstanceIdsInHostGroup(stackV4Response, policyHostGroup);
+
+        int serviceHealthyHostGroupSize = stackResponseUtils.getCloudInstanceIdsWithServicesHealthyForHostGroup(stackV4Response, policyHostGroup).size();
+        int existingHostGroupSize = stackResponseUtils.getCloudInstanceIdsForHostGroup(stackV4Response, policyHostGroup).size();
+
+        int maxResourceValueOffset = loadAlertConfiguration.getMaxResourceValue() - (stopStartEnabled ? serviceHealthyHostGroupSize : existingHostGroupSize);
+        int minResourceValueOffset = serviceHealthyHostGroupSize - loadAlertConfiguration.getMinResourceValue();
+
+        int maxAllowedUpScale = Math.max(maxResourceValueOffset, 0);
+        int maxAllowedDownScale = Math.max(minResourceValueOffset, 0);
+
+        LOGGER.info("Various counts: hostFqdnsToInstanceId={}, servicesHealthyHostFqdnsToInstanceId: {}, stoppedHostFqdnsToInstanceId: {}, " +
+                        "maxResourceValueOffset={}, minResourceValueOffset={}, maxAllowedUpScale={}, maxAllowedDownScale={}",
+                existingHostGroupSize, serviceHealthyHostGroupSize, stoppedHostInstanceIds.size(),
+                maxResourceValueOffset, minResourceValueOffset, maxAllowedUpScale, maxAllowedDownScale);
+
+        Optional<Integer> mandatoryDownScaleCount = Optional.of(maxResourceValueOffset)
+                .filter(mandatoryDownscale -> mandatoryDownscale < 0).map(downscale -> -1 * downscale);
+
+        YarnScalingServiceV1Response yarnResponse =
+                yarnMetricsClient.getYarnMetricsForCluster(cluster, stackV4Response, policyHostGroup, pollingUserCrn, mandatoryDownScaleCount);
+
+        if (cluster.getUpdateFailedDetails() != null && pollingUserCrn.equals(cluster.getMachineUserCrn())) {
+            // Successful YARN API call
+            clusterService.setUpdateFailedDetails(cluster.getId(), null);
+        }
+
+        int yarnRecommendedScaleUpCount = yarnResponseUtils.getYarnRecommendedScaleUpCount(yarnResponse, policyHostGroup, maxAllowedUpScale,
+                loadAlertConfiguration.getMaxScaleUpStepSize());
+        List<String> yarnRecommendedDecommissionHosts =
```

Comment:
Similarly here. maxAllowed, mandatory, stepSize don't need to go to the YARN invocation. Instead compute values here.